apps/yggdrasil/lib/yggdrasil/user.ex

```diff
@@ -29,14 +35,121 @@ defmodule Yggdrasil.User do
   end
 
   @doc """
-  adds password at end of chain protects the hashpwsalt from seeing a nil value
-  in the case of missing password field.
+  Creates the user and assigns the default roles of "player"
+  """
+  def create_with_default_role(attributes) do
+    changeset = User.create_changeset(%User{}, attributes)
+
+    with {:ok, user} <- Repo.insert(changeset),
+          :ok        <- assign_role(user, "player"),
+     do: {:ok, user}
+  end
+
+  def assign_role(user, role_name) do
+    role = Repo.get_by!(Role, name: role_name)
+
+    user_role = UserRole.changeset(%UserRole{},
+                                   %{user_id: user.id, role_id: role.id})
+    case Repo.insert(user_role) do
+      {:ok, _role} -> :ok
+      error        -> error
+    end
+  end
+
+  @doc ~S"""
+  Checks the list of resource/permissions pair given against what the user
+  has been granted and returns true if and only if all permissions supplied
+  have been granted for the user.
+
+  ## roles
+  A user can have one or more roles with each role having a set of resources
+  and a set of resources and those resources having a set of permissions. Roles
+  are not checked directly and only serve as a grouping mechanism for resources.
+
+  Since the user can have multiple roles, the permission set checked for a given
+  resource is the unique set across all the roles for that resource.
+
+  So if the user has `Role1` and `Role2` defined as follows:
+
+  * Role1
+    * foo
+      * :read
+      * :write
+  * Role2
+    * foo
+      * :read
+      * :all
+
+  The set of permissions checked for foo will be `[:read, :write, :all]`
+
+  ## Examples
+      # checking a single resource
+      User.is_granted? user, foo: [:read]
+
+      # checking multiple resources is allowed as well
+      User.is_granted? user, foo: [:read], bar: [:read, :write]
+  """
+  def is_granted?(user, res_perms) do
+    # ensure users roles and the rest are loaded
+    # this does nothing if they are.
+    user = preload_roles(user)
+
+    results = Enum.map res_perms, fn {res, perms} ->
```

Comment:
I can't help but think the original data structure isn't right. Maybe we need to translate the role data into a map of resource to permissions when the user is retrieved. Seems like it would make this far simpler plus minor gains since the user will be cached anyway. Then this function could be more like:

```elixir
def is_granted?(user, res_perms) do
    !Enum.any?(res_perms, fn ({res, perms}) ->
        user.permission
        |> Map.get(res, MapSet.new())
        |> MapSet.disjoint?(perms)
    end)
end
```